Bug3711 - Provide optimistic data by default

sysrqb opened the pull request:
Add support for Torsocks to provide optimistic data, by default, during the SOCKS handshake.

Overload write/read/send/recv/writev/readv/sendmmsg/recvmmsg/recvmsg/recvfrom
Overload pselect/ppoll

sysrqb committed:
Factor the common functionality of select and pselect

When we shim both select and pselect, much of the _guts routines are
identical so we should take advantage of that.

sysrqb committed:
Add pselect functionality equal to select

sysrqb committed:
Factor out the common functionality of poll and ppoll

sysrqb committed:
Add ppoll functionality equal to that of poll

sysrqb committed:
Avoid entering a polling-infinite loop by removing connected sockets

sysrqb committed:
Add body to readv shim

sysrqb committed:
Check if __SOCKADDR_ARG is defined, else use struct sockaddr *

This is a Linux-centric patch, so we'll want to handle this better
later.

Relevant snippets from sys/socket.h:

/* This is the type we use for generic socket address arguments.

   With GCC 2.7 and later, the funky union causes redeclarations or
   uses with any of the listed types to be allowed without complaint.
   G++ 2.7 does not support transparent unions so there we want the
   old-style declaration, too.  */
/* Add more `struct sockaddr_AF' types here as necessary.
   These are all the ones I found on NetBSD and Linux.  */
  __SOCKADDR_ONETYPE (sockaddr) \
  __SOCKADDR_ONETYPE (sockaddr_at) \
  __SOCKADDR_ONETYPE (sockaddr_ax25) \
  __SOCKADDR_ONETYPE (sockaddr_dl) \
  __SOCKADDR_ONETYPE (sockaddr_eon) \
  __SOCKADDR_ONETYPE (sockaddr_in) \
  __SOCKADDR_ONETYPE (sockaddr_in6) \
  __SOCKADDR_ONETYPE (sockaddr_inarp) \
  __SOCKADDR_ONETYPE (sockaddr_ipx) \
  __SOCKADDR_ONETYPE (sockaddr_iso) \
  __SOCKADDR_ONETYPE (sockaddr_ns) \
  __SOCKADDR_ONETYPE (sockaddr_un) \
  __SOCKADDR_ONETYPE (sockaddr_x25)
 ...
typedef union { __SOCKADDR_ALLTYPES
              } __SOCKADDR_ARG __attribute__ ((__transparent_union__));
 ...

sysrqb committed:
Add additional debug messages in select loop

We also need select to tell us when we can read from optdata sockets

sysrqb committed:
Use appropriate errno for read/readv if SOCKS handshake fails

sysrqb committed:
Drop writes by sendmmsg on non-TCP/IP sockets

sysrqb committed:
Check if compiler supports Werror

Enable -Werror with Autoconf when available. Capability checks may
incorrectly pass when a warning is issued.

sysrqb committed:
Correctly verify system's send/recvmmsg capability

The previous check rendered false-positives because we were prototyping
the function. We need to actually call the function before we know for
certain it exists. Only checking the prototype doesn't help us if the
system doesn't have a previous declaration for it.

sysrqb committed:
Add definition and placeholder for new routines

sysrqb committed:
Catch unknown symbols before calling

sysrqb committed:
Check for syscalls in a way that actually works

sysrqb committed:
Add new configure checks and preprocessor definitions

We need to intercept additional functions if we want to be able to
handle the clients correctly.

Check for recvmsg, read, write, send, recv, recvfrom, writev, readv,
recvmmsg, sendmmsg, ppoll, pselect

sysrqb committed:
Define ppoll signature as void when unsupported

Conflicts:
	src/torsocks.c

sysrqb committed:
Add additional debug messages

Conflicts:
	src/torsocks.c

sysrqb committed:
Block-out test that fails due to -Werror

sysrqb committed:
Basic framework for supporting optimistic data

Enable optimistic data by default, but allow for it to be disabled in
the config file.

sysrqb committed:
Provide minimal shimming of I/O

Add minimal interference for wrapping write/read/send/recv/writev/readv

sysrqb committed:
Fix clang warning

Extraneous parenthesis

sysrqb committed:
Fix issues found by dgoulet

Remove select_common and poll_common from torsocks_ namespace
Make sure opts->sigmask is initialized

sysrqb committed:
Merge branch 'bug3711_1' of https://github.com/sysrqb/torsocks into bug3711_1

dgoulet commented:
This looks good in terms of code review but it seems there is a regression going on from master to this branch which is that I'm unable to perform a simple "wget http://google.com" with this branch.

Here is the strace output where the first one is what it should be from master and the second one is this branch where is stalls until an http timeout.

8231  sendto(3, "\4\1\0P\0\0\0\1dgoulet\0google.com\0", 27, 0, NULL, 0) = 27
8231  recvfrom(3, "\0Z\0\0\0\0\0\0", 8, 0, NULL, NULL) = 8
8231  write(2, "connected.\n", 11)      = 11
8231  select(4, NULL, [3], NULL, {900, 0}) = 1 (out [3], left {899, 999998})
8231  write(3, "GET / HTTP/1.1\r\nUser-Agent: Wget"..., 110) = 110
[...]

This branch:

11711 sendto(3, "\4\1\0P\0\0\0\1dgoulet\0google.com\0", 27, 0, NULL, 0) = 27
11711 write(2, "connected.\n", 11)      = 11
11711 select(4, [3], [], [3], {900, 0}) = 1 (in [3], left {899, 445663})
11711 recvfrom(3, "\0Z\0\0\0\0\0\0", 8, 0, NULL, NULL) = 8
11711 select(4, [3], [], [], {899, 445663}) = ? ERESTARTNOHAND (To be restarted)
<...stalled>

sysrqb commented:
Wow. hm. strange. I guess somehow I never tested http, only https (really bad assumption). Working on a fix now.

sysrqb committed:
Select must explicitly set read/write for optdata

When requesting optimistic data from an application while in a select
loop, explicitly reset read/write events, else we'll think they were for
us regarding the handshake and not for the application.

sysrqb commented:
Ok, moderately bad hack, but please test/review and I'll adjust as is needed.
